### DISCUSSION TITLE
How to convert a TIFF  GIS file to be usable in GIMP and Blender as a heightmap

### DISCUSSION THREAD

**Discussion by Moonwards** [ImageMagick/ImageMagick#4325] [Help] [answered]
I downloaded a large file meant for use in GIS software and want to use it to make a heightmap for use in Blender. If I open it in GIMP, it looks just black and white. I vaguely recall having had the same problem a few years ago in a very similar situation, and it turned out the data was all there, it just wasn't being read properly. The grayscale had to be converted somehow. To a 16-bit PNG, I guess, but the color data also needs something, I believe.

I cut out the area I need in GIMP and saved it as a new file of much more manageable size,. guessing that such an action wouldn't be a problem.

Is there anyone here with experience in these kinds of files who knows how I should proceed?

Here is a small screenshot of how it looks right now. If I didn't know better, I'd say it's two tones - black, and white.
![Screenshot from 2021-10-06 18-42-27](https://user-images.githubusercontent.com/50932660/136304927-dad001d9-df85-4e86-999b-33b4c7c9ec7b.png)

**Comment by fmw42** [comment]
The issue might be that you have a lower dynamic range of values, something like 10 or 12 or 14-bit data, that is put into a 16-bit file. So the image will look dark, because no gray level is as bright as full 16-bits. You can scale the result if you know the image range by multiplying or you can just auto-level it.  The first thing would be to get the image statistics.

`identify -verbose image`

if on IM 6 and on IM 7

`magick identify -verbose image`

Alternately, you can just convert it with an automatic stretch of dynamic range

convert image -auto-level new_image

or 

`magick image -auto-level new_image`

**Reply by Moonwards** [reply]
The terminal didn't recognize the `magick` command, though I have 6.9.10, on Ubuntu 20.04. 
However I did something similar with exiftool, and the problem seems to be the opposite. It's a 32 bit image.

**Comment by fmw42** [comment]
Use convert not magick since you have 6.9.10 on linux.  Or identify, not magick identify.  If you have less than 32-bits in the file, then the same issue. So we need to know what the actual range of gray level in the image would be, not the bit-depth of the file.

`identify -verbose image.tiff`

be sure image.tiff is in your working (home) directory if that is where you are using the terminal,

With ImageMagick and 32-bit tif, you might have to use

`identify -verbose -define quantum:format=floating-point image.tiff`

or

`convert image.tiff -define quantum:format=floating-point -verbose info:`

**Reply by fmw42** [reply]
Also what do you get from 

`convert -version`

**Reply by Moonwards** [reply]
Ah, ok.
I get 
```
Depth: 32/16-bit
  Channel depth:
    gray: 16-bit
  Channel statistics:
    Pixels: 82575360
    Gray:
      min: 0  (0)
      max: 65535 (1)
      mean: 1945.88 (0.0296922)
      standard deviation: 11119.1 (0.169666)
      kurtosis: 28.715
      skewness: 5.54159
      entropy: 0.0146491

```
And for convert -version
`Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101`

**Comment by fmw42** [comment]
So an unsigned integer is a 32-bit datum that encodes a nonnegative integer in the range 0 to (2**32)-1 or [0 to 4294967295] and a 16-bit datum is in the range 0 to (2**16)-1 or [0 to 65535]. So divide 4294967295/65535=65537. So you can do

`convert image.tif -define quantum:format=floating-point -evaluate multiply 65537 new_image.tif`

Or as I said above stretch to whatever actual dynamic range you have automatically by

`convert image -define quantum:format=floating-point -auto-level new_image.tif`

**Reply by Moonwards** [reply]
Nope, both commands output a file that looks identical.

I got advice elsewhere to use a tool GDAL. It' computed the min and max values in the file, and output -3184.302, 774.906
When I tried to use that tool for conversion after that step it threw and error. 
I tried there to remove the negative number by rounding and making the min 0 and the max 775 + 3184. That didn't work either. Could the negative number be an issue?

**Comment by snibgo** [comment]
@Moonwards: Please post, or link to, a file that works as "heightmap for use in Blender". Please also post a file that you want to convert to that format.

**Comment by fmw42** [comment]
You apparently have a signed 32-bit file. I suspect you will need be using an HDRI compiled version of ImageMagick. Yours is not HDRI.  The default ImageMagick 6 is not HDRI. But the default ImageMagick 7 is HDRI.

Perhaps you can post your image or if too big extract a small section of your image and post that. Use your other tools that can read it and crop it (GIMP?) and zip it for safety. Then post that image.

**Comment by Moonwards** [comment]
Here is a file from NASA, it comes from [the LROC archive](http://wms.lroc.asu.edu/lroc/rdr_product_select)

[Lalande_Raw_Heightmap_WAC_GLD100_100M_512.zip](https://github.com/ImageMagick/ImageMagick/files/7298530/Lalande_Raw_Heightmap_WAC_GLD100_100M_512.zip)

Here is a snippet of the file I'm trying to convert, it comes from [China's Lunar and Planetary Data Release System](https://moon.bao.ac.cn/ce5web/searchOrder_dataSearchData.search)

[Chang2-TestArea.zip](https://github.com/ImageMagick/ImageMagick/files/7298535/Chang2-TestArea.zip)

I have to call it a day. I'll get ImageMagick 7 in the morning.

**Comment by fmw42** [comment, marked as answer]
ImageMagick 7 with HDRI shows your data to be:

`magick -quiet Chang2-TestArea.tif -define quantum:format=floating-point -verbose info:`

```
 Depth: 32/16-bit
  Channel depth:
    Gray: 16-bit
  Channel statistics:
    Pixels: 1507328
    Gray:
      min: -1.7736e+08  (-2706.35)
      max: 3.07351e+07 (468.988)

```
So the minimum value is -2706.34 and the maximum is 468.988.

I need to convert to 16-bit integer, since I am on a 16-bit Q16 HDRI ImageMagick, not 32-but Q32 HDRI. So I can do one of two things.

What do you want to do with the negative values? 

1) divide by 469 (your max value) and discard negatives

`magick -quiet Chang2-TestArea.tif -define quantum:format=floating-point -evaluate divide 469 -clamp  Chang2-TestArea_scaled.tif 
`

[Chang2-TestArea_scaled.tif.zip](https://github.com/ImageMagick/ImageMagick/files/7298738/Chang2-TestArea_scaled.tif.zip)


or

2) Linearly Stretch -2706 to 0 and 468 to 65535. This can be done automatically with -auto-level

`magick -quiet Chang2-TestArea.tif -define quantum:format=floating-point -auto-level Chang2-TestArea_al.tif `

[Chang2-TestArea_al.tif.zip](https://github.com/ImageMagick/ImageMagick/files/7298740/Chang2-TestArea_al.tif.zip)

The first one is more accurate, since it is just a multiply factor.

The second one looks better since if fills the full dynamic range, but the scaling relationship is unknown but could be computed and would be a multiply and an add.

**Reply by Moonwards** [reply]
Okay, I've gotten the output of both commands, it worked great. Those min and max value look spot on, so I'm going to take the ones I got for the full image and use them to establish the correct height of the mesh made from this. I'll try with both to see what difference it makes and post a shot of the final outcome.
Thanks for your help!

**Comment by snibgo** [comment]
The file Lalande_Raw_Heightmap_WAC_GLD100_100M_512.tiff is grayscale, one channel, integer, 8 bits per channel. So I assume that's the format you want. It has values between 78 and 130 (out of 255), with a peak around 107.

The file Chang2-TestArea.tif is grayscale, one channel, floating-point, with values between -2706.35 and +468.988 as multiples of QuantumRange. This is a massive range. On this scale, black is 0.0 and white is 1.0. So most of the values are either blacker than black, or whiter than white. What do those values represent? I expect the linked web page explains somewhere, but I can't read Chinese.

A possible conversion is:
```
magick Chang2-TestArea.tif -auto-level -define quantum:format=integer -depth 8 out.tif
```
The output is grayscale, one channel, integer, 8 bits per channel, as required.

As fmw42 says, the scale from this is probably wrong.

**Reply by Moonwards** [reply]
The GLD100_512 file is 8 bits? Yeah, I guess that makes sense - I didn't think about it. But I went and found the other file from the Chinese space agency because it has a lot more detail, and I don't want to lose that. I want to keep the 16 bit. Sorry for not being clear, I've only ever worked with heightmaps for this one specific crater and its region. I struggle with it each time it comes up.

I'm going to make the effort to install ImageMagick 7 and try the commands fmw42 says. The value range actually looks very right - it's saying the bottom of the crater is -2706 m below lunar 'sea level', and the top of the rim is 469 m above it. I know that crater, and that is spot on. Thanks for going to the trouble to tell me how to do it with an 8 bit output, though.

**Comment by snibgo** [comment]
The documentation for Blender, or whatever software reads the DEM files, should say what formats it accepts, and how to represents heights such as -2706 metres and +469 metres. Perhaps the format needs to be signed integer, or unsigned integer, or perhaps it can accept floating-point.

Once you know what formats are accepted, we can show what IM options to use.

**Reply by Moonwards** [reply]
Well, it might say that somewhere in the documentation, but this isn't typically how Blender is used. The setup to turn this into a landscape does use the information in a relevant way. 
I can add the absolute values of the minimum and maximum and put that in as the value in the `Strength` field in the Displacement modifier used to generate the model. And I can divide the Max value by the Min value to get a ratio that goes in the Mid-level field. 
That sets it up properly, in the case of this dataset.
